docs/how-packages-and-artifacts-are-built.md

```diff
@@ -0,0 +1,114 @@
+
+The purpose of the above configuration is to generate two bootstrap tarballs with each build. The default tarball will contain the code that handles the import of the remaining DC/OS packages (as dictated by the provider `gen.build_deploy` module) and setup the hosts to startup DC/OS. The installer tarball will be used to create a mock DC/OS environment which can provide packages to hosts via the `dcos_installer` program (see below).
+
+#### buildinfo.json
+Variants can also allow for customized packages. In any given package, one may add a `foobar.buildinfo.json` that will pull in alternative sources to run against the same `build` script. To have a package variant built, the `treeinfo.json` must have a `"variant"` section that describes the specific package variant to be used with the tree variant. This logic must be done explicitly as there is no implicit matching between package and tree variants.
+
+#### upstream.json
+The other file that can be placed in the `packages` directory is called `upstream.json`. This JSON is formatted like a single source from the aforementioned `buildinfo.json`. Here's an example:
+```
+{
+  "git": "https://github.com/dcos/dcos",
+  "kind": "git",
+  "ref": "9364e0a72b942e8dcec306ef7c1673c6f23bd9cc",
+  "ref_origin": "master"
+}
+```
+This file will tell pkgpanda the packages directory containing it should be expanded to include the packages from the given upstream source. Thus, one may create a modified version of DC/OS by just defining the desired package modifications; no forked repositories required!
+
+### Developing Packages
+When building a release, `pkgpanda` is only called through its library. However, there is a CLI which is very helpful for developing and debugging packages. To setup pkgpanda locally, be sure docker is installed and able to run without sudo, then run the `./prep_local` script in the root directory (python virtualenv will be required). Finally, make sure the dcos-builder image has been prepped locally by doing `cd pkgpanda/build/docker/dcos-builder/; docker build -t dcos/dcos-builder:dcos-builder_dockerdir-latest .` Right now, this is handled apart of the release scripting, not pkgpanda.
+
+After `./prep_local` finishes, the `mkpanda` command will be available to use. With this, you can navigate to a specific package directory and simply run `mkpanda`. This will instruct pkgpanda to look at the given package, evaluate its dependency list, build those if necessary, and then finally build the chosen package.
+
+From the `packages` directory, one can run `mkpanda tree` which will essentially do a full, locally-cached DC/OS build. Alternatively, one can name a variant tree like so: `mkpanda tree installer`. This will instruct pkgpanda to only make the packages necessary for building the completed variant.
+
+## Deployment Artifacts
+The artifacts are actually delivered to hosts via one of the deployment methods crafted in the `do_create` function of the modules in `gen.build_deploy`.
```

Comment:
You could consider avoiding the mention of these method names like `do_create`. If we refactor the code and miss out this document, it could confuse people.